src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Definition/RazorDefinitionEndpoint.cs

```diff
@@ -0,0 +1,154 @@
+        {
+            return new TextDocumentRegistrationOptions
+            {
+                DocumentSelector = RazorDefaults.Selector
+            };
+        }
+
+        public async Task<LocationOrLocationLinks> Handle(DefinitionParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                var path = request.TextDocument.Uri.GetAbsoluteOrUNCPath();
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(documentSnapshot.FileKind))
+            {
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var originTagHelperBinding = await GetOriginTagHelperBindingAsync(documentSnapshot, codeDocument, request.Position).ConfigureAwait(false);
+            if (originTagHelperBinding is null)
+            {
+                return null;
+            }
+
+            var originTagDescriptor = originTagHelperBinding.Descriptors.SingleOrDefault();
+            if (originTagDescriptor is null)
+            {
+                return null;
+            }
+
+            var originComponentDocumentSnapshot = await _componentSearchEngine.TryLocateComponentAsync(originTagDescriptor).ConfigureAwait(false);
+            if (originComponentDocumentSnapshot is null)
+            {
+                return null;
+            }
+
+            var originComponentUri = new UriBuilder
+            {
+                Path = originComponentDocumentSnapshot.FilePath,
+                Scheme = Uri.UriSchemeFile,
+                Host = string.Empty,
+            }.Uri;
+
+            return new LocationOrLocationLinks(new[]
+            {
+                new LocationOrLocationLink(new Location
+                {
+                    Uri = originComponentUri,
+                    Range = new Range(new Position(0, 0), new Position(0, 0)),
+                }),
+            });
+        }
+
+        public void SetCapability(DefinitionCapability capability)
+        {
+            _capability = capability;
+        }
+
+        private async Task<TagHelperBinding> GetOriginTagHelperBindingAsync(DocumentSnapshot documentSnapshot, RazorCodeDocument codeDocument, Position position)
+        {
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
```

Comment:
If you pull this entire section out (line 117 - line 120) the method could then become `GetOriginTagHelperBindingAsync(RazorCodeDocument codeDocument, SourceLocation location)`

At which point you could make this internal static and add unit tests.